language/evm/move-to-yul/src/functions.rs

```diff
@@ -107,48 +108,260 @@ impl<'a> FunctionGenerator<'a> {
-                if let BlockContent::Basic { lower, upper } = cfg.content(entry_bb) {
-                    for offs in *lower..*upper + 1 {
-                        self.bytecode(ctx, fun_id, target, &label_map, &code[offs as usize], false);
+
+            if !ctx.options.apply_cfg_to_scf() {
+                self.emit_cfg(&cfg, code, ctx, fun_id, target);
+            } else {
+                self.emit_scf_from_cfg(&cfg, code, ctx, fun_id, target);
+            }
+        });
+        emitln!(ctx.writer)
+    }
+
+    fn emit_cfg(
+        &mut self,
+        cfg: &StacklessControlFlowGraph,
+        code: &[Bytecode],
+        ctx: &Context,
+        fun_id: &QualifiedInstId<FunId>,
+        target: &FunctionTarget,
+    ) {
+        let entry_bb = Self::get_actual_entry_block(cfg);
+        let label_map = Self::compute_label_map(cfg, code);
+        // Emit state machine to represent control flow.
+        // TODO: Eliminate the need for this, see also
+        //    https://medium.com/leaningtech/solving-the-structured-control-flow-problem-once-and-for-all-5123117b1ee2
+        if cfg.successors(entry_bb).iter().all(|b| cfg.is_dummmy(*b)) {
+            // In this trivial case, we have only one block and can omit the state machine
+            if let BlockContent::Basic { lower, upper } = cfg.content(entry_bb) {
+                for offs in *lower..*upper + 1 {
+                    self.bytecode(
+                        ctx,
+                        fun_id,
+                        target,
+                        &label_map,
+                        &code[offs as usize],
+                        false,
+                        false,
+                    );
+                }
+            } else {
+                panic!("effective entry block is not basic")
+            }
+        } else {
+            emitln!(ctx.writer, "let $block := {}", entry_bb);
+            emit!(ctx.writer, "for {} true {} ");
+            ctx.emit_block(|| {
+                emitln!(ctx.writer, "switch $block");
+                for blk_id in &cfg.blocks() {
+                    if let BlockContent::Basic { lower, upper } = cfg.content(*blk_id) {
+                        // Emit code for this basic block.
+                        emit!(ctx.writer, "case {} ", blk_id);
+                        ctx.emit_block(|| {
+                            for offs in *lower..*upper + 1 {
+                                self.bytecode(
+                                    ctx,
+                                    fun_id,
+                                    target,
+                                    &label_map,
+                                    &code[offs as usize],
+                                    true,
+                                    false,
+                                );
+                            }
+                        })
                     }
+                }
+            })
+        }
+    }
+
+    fn emit_scf_from_cfg(
```

Comment:
As we discussed offline, the construction of the SCG really doesn't belong in this module. It should fully go into the module which defines this data type.